jvm/workbookapp/src/test/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/persistence/dao/AudioPluginDaoTest.kt

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (C) 2020-2022 Wycliffe Associates
+ *
+ * This file is part of Orature.
+ *
+ * Orature is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Orature is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Orature.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package org.wycliffeassociates.otter.jvm.workbookapp.persistence.dao
+
+import org.jooq.exception.DataAccessException
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+import org.wycliffeassociates.otter.jvm.workbookapp.persistence.database.AppDatabase
+import org.wycliffeassociates.otter.jvm.workbookapp.persistence.database.InsertionException
+import org.wycliffeassociates.otter.jvm.workbookapp.persistence.entities.AudioPluginEntity
+import java.io.File
+
+class AudioPluginDaoTest {
+    private val testDatabaseFile = File.createTempFile(
+        "test-audio-plugin-dao", ".sqlite"
+    ).also(File::deleteOnExit)
+    private lateinit var database: AppDatabase
+    private val dao by lazy { database.audioPluginDao }
+
+    private val defaultEnity = AudioPluginEntity(
+        id = 0,
+        name = "test",
+        version = "test",
+        bin = "test-jar",
+        args = "test",
+        edit = 0,
+        record = 0,
+        mark = 0,
+        path = null
+    )
+
+    @Before
+    fun setup() {
+        database = AppDatabase(testDatabaseFile)
+    }
+
+    @Test
+    fun testInsert() {
+        Assert.assertEquals(0, dao.fetchAll().size)
+        val id1 = dao.insert(defaultEnity)
+        Assert.assertEquals(1, dao.fetchAll().size)
+
+        // insert the same entity will trigger an update
+        val id2 = dao.insert(defaultEnity)
+
+        Assert.assertEquals(1, dao.fetchAll().size)
+        Assert.assertEquals(id1, id2)
+    }
+
+    @Test
+    fun testInsertThrowsException() {
+        try {
+            dao.insert(defaultEnity.copy(id = 1))
+            Assert.fail("Inserting nonzero id should throw an exception")
```

Comment:
Good use of failure message.  It's a good idea to use messages on other asserts too, e.g. `assertEquals(1, dao.fetchAll().size, "After inserting, there should be one record.")`